### DISCUSSION TITLE
Moving to ESM

### DISCUSSION THREAD

**Discussion by lifeiscontent** [sindresorhus/awesome-nodejs#1093] [Ideas]
Might be interesting to have an "examples" directory to showcase the proper ways of creating and maintaining packages with a variety of build tools such as Babel, typescript, esbuild, nothing at all.

That way people can read and learn from good code examples.

**Comment by sindresorhus** [comment]
It's not something I want to maintain, but someone could make a repo for this and submit a link to it here.

I personally prefer no build tooling at all.

**Reply by lifeiscontent** [reply]
@sindresorhus do you have any packages written in TypeScript?

The process of publishing a package in typescript is confusing to me; for example, if you have files that you want to reference in the build e.g. my-lib/utils -> src/utils.ts do you know how to deal with these things?

**Reply by sindresorhus** [reply]
Yes:
- https://github.com/sindresorhus/got
- https://github.com/sindresorhus/p-queue
- https://github.com/sindresorhus/ow

**Reply by papb** [reply]
@lifeiscontent

> if you have files that you want to reference in the build

Can you clarify your confusion? I can try to help you

**Reply by lifeiscontent** [reply]
@papb its not that I'm confused more so that if I have a typescript project that looks like the following:

```
.
├── package.json
├── src
│   ├── index.ts
│   └── utils.ts
├── tsconfig.json
└── yarn.lock
```

how do I then reference index and utils as so:

`import foo from 'my-package'`

and

`import bar from 'my-package/utils'`

I know I could stick a utils.js in the root of my project to reference the `dist` folder, but that seems like a hack and I would expect there to be utils for packaging precompiled projects.

**Reply by papb** [reply]
Ah, I see. Actually I just stumbled upon the same thing last week. I don't know either :sweat_smile:. Amazing coincidence! If you figure it out let me know. I'll do the same. For now, I am doing precisely what you called a hack (and I agree with you)

Let's continue talking about this on https://github.com/sindresorhus/project-ideas/issues/130, since this got off-topic here.

**Comment by lifeiscontent** [comment]
@sindresorhus thank you so much for the info, it was super helpful. I've converted my entire monorepo to follow this format thanks to your examples. ❤️

**Comment by Joooojj** [comment]
![image](https://github.com/sindresorhus/awesome-nodejs/assets/160371587/465928ab-22ec-4500-91e4-3c53a05bb0f1)
